packages/flutter/lib/src/cupertino/tab_scaffold.dart

```diff
@@ -156,29 +245,54 @@ class CupertinoTabScaffold extends StatefulWidget {
 }
 
 class _CupertinoTabScaffoldState extends State<CupertinoTabScaffold> {
-  int _currentPage;
+  CupertinoTabController _controller;
 
   @override
   void initState() {
     super.initState();
-    _currentPage = widget.tabBar.currentIndex;
+    _updateTabController();
+  }
+
+  void _updateTabController() {
+    final CupertinoTabController newController =
+      // User provided a new controller, update [_controller] with it.
+      widget.controller
+      // Always create this [_CupertinoTabScaffoldState]'s own controller,
+      // if oldWidget.controller != null && widget.controller == null
+      ?? CupertinoTabController(initialIndex: widget.tabBar.currentIndex);
+
+    if (newController == _controller) {
+      return;
+    }
+
+    _controller?.removeListener(_onCurrentIndexChange);
+    newController.addListener(_onCurrentIndexChange);
+    _controller = newController;
+  }
+
+  void _onCurrentIndexChange() {
+    assert(() {
+        if (_controller.index < 0 || _controller.index >= widget.tabBar.items.length) {
+          throw FlutterError(
+            "The CupertinoTabController's current index ${_controller.index} is"
+            'out of bounds for the tab bar with ${widget.tabBar.items.length} tabs'
+          );
+        }
+        return true;
+      }());
+
+    setState(() {});
```

Comment:
Don't leave the body empty. Add a code comment saying what would read from _controller.value after.